File: crates/taflow-core/src/indicators/mc_clellan_summation_index.rs
use crate::error::TaResult;
use crate::indicators::McClellanOscillator;
/// Running sum of the classic McClellan Oscillator.
#[derive(Debug, Clone)]
pub struct McClellanSummationIndex {
    oscillator: McClellanOscillator,
    total: f64,
    count: usize,
    value: Option<f64>,
}
impl McClellanSummationIndex {
    /// Create an empty oscillator and cumulative total.
    pub fn new() -> TaResult<Self> {
        Ok(Self {
            oscillator: McClellanOscillator::new()?,
            total: 0.0,
            count: 0,
            value: None,
        })
    }
    /// Append aggregate advancing and declining issue counts.
    pub fn append(&mut self, advancers: f64, decliners: f64) -> Option<f64> {
        self.count += 1;
        if let Some(x) = self.oscillator.append(advancers, decliners) {
            self.total += x;
            self.value = Some(self.total);
        }
        self.value
    }
    /// Return the latest cumulative index value.
    pub fn value(&self) -> Option<f64> {
        self.value
    }
    /// Return the number of processed market ticks.
    pub fn len(&self) -> usize {
        self.count
    }
    /// Return whether no market ticks have been processed.
    pub fn is_empty(&self) -> bool {
        self.count == 0
    }
    /// Reset the embedded oscillator and cumulative total.
    pub fn reset(&mut self) {
        self.oscillator.reset();
        self.total = 0.0;
        self.count = 0;
        self.value = None;
    }
}

File: python/taflow/indicators/mc_clellan_summation_index.py
"""Python adapter for the native McClellan Summation Index."""

from typing import Any
import numpy as np
from .._native import McClellanSummationIndex as _Native
from .._series import as_float64_series


class McClellanSummationIndex:
    """Accumulate the McClellan Oscillator across market breadth ticks.

    Native Rust computes the Wickra-compatible 19/39 oscillator from
    ratio-adjusted net advances and cumulatively sums every emitted value. This
    maps to Wickra ``McClellanSummationIndex``.

    Args:
        advancers: Number of advancing issues at each tick.
        decliners: Number of declining issues at each aligned tick.

    Raises:
        ValueError: If the two histories have different lengths.
    """

    def __init__(self) -> None:
        """Initialize an empty configured native state.
        """
        self._state = _Native()

    def append(
        self, advancers: float, decliners: float
    ) -> "McClellanSummationIndex":
        """Append one breadth tick and return this adapter."""
        self._state.append(float(advancers), float(decliners))
        return self

    def extend(self, advancers: Any, decliners: Any) -> "McClellanSummationIndex":
        """Append aligned breadth histories after length validation."""
        arrays = as_float64_series(advancers), as_float64_series(decliners)
        if len(arrays[0]) != len(arrays[1]):
            raise ValueError("advancers and decliners must have equal lengths")
        self._state.extend(*arrays)
        return self

    @property
    def value(self) -> float | None:
        """Return the latest cumulative index, or ``None`` before input."""
        return self._state.value

    def compute(self) -> np.ndarray:
        """Return the aligned McClellan Summation Index history."""
        return self._state.compute()

    def reset(self) -> "McClellanSummationIndex":
        """Reset oscillator and cumulative state and return this adapter."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the processed-tick count delegated to native state."""
        return len(self._state)
